### PR TITLE
APIs for creation of disabled participants <master> [7887]

### PR DESCRIPTION
**Pull request by IkerLuengo**
This PR solves issues with DDSSimpleCommunicationTypeDiscovery on master.

It also reworks #1044 using the new solution, as the original one may cause a mutex interlock.

### CODE REVIEW

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp [hunk 1]
```diff
@@ -307,10 +318,13 @@ RTPSParticipantImpl::RTPSParticipantImpl(
...
-    for (auto& receiver : m_receiverResourcelist)
+    if (enabled)
     {
-        receiver.Receiver->RegisterReceiver(receiver.mp_receiver);
+        //Start reception
+        for (auto& receiver : m_receiverResourcelist)
+        {
+            receiver.Receiver->RegisterReceiver(receiver.mp_receiver);
+        }
     }
```
Comment (MiguelCompany) [inline src/cpp/rtps/participant/RTPSParticipantImpl.cpp]: This whole block can be removed, as it is done inside `enable()`. This also means that we can remove the `enabled` argument

File: src/cpp/rtps/participant/RTPSParticipantImpl.h [hunk 2]
```diff
@@ -159,18 +159,48 @@ class RTPSParticipantImpl
             RTPSParticipant* part,
             RTPSParticipantListener* plisten = nullptr);
 
+    /**
```
Comment (MiguelCompany) [inline src/cpp/rtps/participant/RTPSParticipantImpl.h]: Whole changes on this file could be discarded

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1545/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1317/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1543/)

**Comment by MiguelCompany** [comment]
I could catch an ABBA lock on the subscriber while deleting the participant

```
Reception thread
----------------

    Mutex order: 
    * MessageReceiver::mtx_ taken on MessageReceiver::proc_Submsg_Data
    * StatelessReader::mp_mutex taken on StatelessReader::processDataMsg
    * Blocked trying to take SubscriberImpl::mtx_readers_ on SubscriberImpl::lookup_datareader (called from main SubListener::on_new_data_message in test/dds/communication/Subscriber.cpp)

 	[Código externo]	
 	[Marco flotante] fastrtps-1.10.dll!std::_Mtx_lockX(_Mtx_internal_imp_t *) Línea 72	C++
 	[Marco flotante] fastrtps-1.10.dll!std::_Mutex_base::lock() Línea 49	C++
 	[Marco flotante] fastrtps-1.10.dll!std::lock_guard<std::mutex>::{ctor}(std::mutex &) Línea 517	C++
 	fastrtps-1.10.dll!eprosima::fastdds::dds::SubscriberImpl::lookup_datareader(const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & topic_name) Línea 245	C++
 	DDSSimpleCommunicationSubscriber.exe!SubListener::on_new_data_message(eprosima::fastdds::dds::Subscriber * subscriber) Línea 203	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::StatelessReader::change_received(eprosima::fastrtps::rtps::CacheChange_t * change) Línea 168	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::StatelessReader::processDataMsg(eprosima::fastrtps::rtps::CacheChange_t * change) Línea 305	C++
 	[Marco flotante] fastrtps-1.10.dll!eprosima::fastrtps::rtps::MessageReceiver::proc_Submsg_Data::__l2::<lambda_ddba44a2b4f86fab1988d1a4e605ba31>::operator()(eprosima::fastrtps::rtps::RTPSReader * reader) Línea 666	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::MessageReceiver::findAllReaders<<lambda_ddba44a2b4f86fab1988d1a4e605ba31> >(const eprosima::fastrtps::rtps::EntityId_t & readerID, const eprosima::fastrtps::rtps::MessageReceiver::proc_Submsg_Data::__l2::<lambda_ddba44a2b4f86fab1988d1a4e605ba31> & callback) Línea 504	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::MessageReceiver::proc_Submsg_Data(eprosima::fastrtps::rtps::CDRMessage_t * msg, eprosima::fastrtps::rtps::SubmessageHeader_t * smh) Línea 670	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::MessageReceiver::processCDRMsg(const eprosima::fastrtps::rtps::Locator_t & loc, eprosima::fastrtps::rtps::CDRMessage_t * msg) Línea 251	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::ReceiverResource::OnDataReceived(const unsigned char * data, const unsigned int size, const eprosima::fastrtps::rtps::Locator_t & localLocator, const eprosima::fastrtps::rtps::Locator_t & remoteLocator) Línea 100	C++
 	fastrtps-1.10.dll!eprosima::fastdds::rtps::UDPChannelResource::perform_listen_operation(eprosima::fastrtps::rtps::Locator_t input_locator) Línea 68	C++
 	[Código externo]	

Main thread
-----------

    Mutex order: 
    * SubscriberImpl::mtx_readers_ taken on SubscriberImpl::~SubscriberImpl()
    * Blocked trying to take MessageReceiver::mtx_ on MessageReceiver::removeEndpoint

 	[Código externo]	
 	[Marco flotante] fastrtps-1.10.dll!std::_Mtx_lockX(_Mtx_internal_imp_t *) Línea 72	C++
 	[Marco flotante] fastrtps-1.10.dll!std::_Mutex_base::lock() Línea 49	C++
 	[Marco flotante] fastrtps-1.10.dll!std::lock_guard<std::mutex>::{ctor}(std::mutex &) Línea 517	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::MessageReceiver::removeEndpoint(eprosima::fastrtps::rtps::Endpoint * to_remove) Línea 114	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::RTPSParticipantImpl::deleteUserEndpoint(eprosima::fastrtps::rtps::Endpoint * p_endpoint) Línea 1008	C++
 	fastrtps-1.10.dll!eprosima::fastrtps::rtps::RTPSDomain::removeRTPSReader(eprosima::fastrtps::rtps::RTPSReader * reader) Línea 340	C++
 	fastrtps-1.10.dll!eprosima::fastdds::dds::DataReaderImpl::~DataReaderImpl() Línea 117	C++
 	[Código externo]	
 	fastrtps-1.10.dll!eprosima::fastdds::dds::SubscriberImpl::~SubscriberImpl() Línea 78	C++
 	[Código externo]	
 	fastrtps-1.10.dll!eprosima::fastdds::dds::DomainParticipantImpl::~DomainParticipantImpl() Línea 133	C++
 	[Código externo]	
 	fastrtps-1.10.dll!eprosima::fastdds::dds::DomainParticipantFactory::delete_participant(eprosima::fastdds::dds::DomainParticipant * part) Línea 148	C++
 	DDSSimpleCommunicationSubscriber.exe!main(int argc, char * * argv) Línea 369	C++
 	[Código externo]	

```

**Comment by MiguelCompany** [comment]
It seems we are not correctly following the DDS spec:

```
2.2.2.2.2.2 delete_participant
This operation deletes an existing DomainParticipant. This operation can only be invoked if all domain entities belonging to
the participant have already been deleted. Otherwise the error PRECONDITION_NOT_MET is returned.
```

**Comment by JLBuenoLopez-eProsima** [comment]
Closing due to inactivity. It can be revisited in the future.

**Pull request closed without merging**